template.Debian/docker-entrypoint.sh

```diff
@@ -0,0 +1,238 @@
+			wantHelp=1
+			break
+			;;
+	esac
+done
+
+# usage: file_env VAR [DEFAULT]
+#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
+# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
+#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
+file_env() {
+	local var="$1"
+	local fileVar="${var}_FILE"
+	local def="${2:-}"
+	if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
+		echo >&2 "$(date --rfc-3339=seconds) [ERROR] [Entrypoint]: Both $var and $fileVar are set (but are exclusive)"
+		exit 1
+	fi
+	local val="$def"
+	if [ "${!var:-}" ]; then
+		val="${!var}"
+	elif [ "${!fileVar:-}" ]; then
+		val="$(< "${!fileVar}")"
+	fi
+	export "$var"="$val"
+	unset "$fileVar"
+}
+
+# usage: process_init_file FILENAME MYSQLCOMMAND...
+#    ie: process_init_file foo.sh mysql -uroot
+# (process a single initializer file, based on its extension. we define this
+# function here, so that initializer scripts (*.sh) can use the same logic,
+# potentially recursively, or override the logic used in subsequent calls)
+process_init_file() {
+	local f="$1"; shift
+	local mysql=( "$@" )
+
+	case "$f" in
+		*.sh)     echo "$(date --rfc-3339=seconds) [Note] [Entrypoint]: $0: running $f"; . "$f" ;;
+		*.sql)    echo "$(date --rfc-3339=seconds) [Note] [Entrypoint]: $0: running $f"; "${mysql[@]}" < "$f"; echo ;;
+		*.sql.gz) echo "$(date --rfc-3339=seconds) [Note] [Entrypoint]: $0: running $f"; gunzip -c "$f" | "${mysql[@]}"; echo ;;
+		*)        echo "$(date --rfc-3339=seconds) [Note] [Entrypoint]: $0: ignoring $f" ;;
+	esac
+	echo
+}
+
+_check_config() {
+	toRun=( "$@" --verbose --help )
+	if ! errors="$("${toRun[@]}" 2>&1 >/dev/null)"; then
+		datestring=$(date --rfc-3339=seconds)
+		cat >&2 <<-EOM
+
+			$datestring [ERROR] [Entrypoint]: mysqld failed while attempting to check config
+			command was: "${toRun[*]}"
+
+			$errors
+		EOM
+		exit 1
+	fi
+}
+
+# Fetch value from server config
+# We use mysqld --verbose --help instead of my_print_defaults because the
+# latter only show values present in config files, and not server defaults
+_get_config() {
+	local conf="$1"; shift
+	"$@" --verbose --help --log-bin-index="$(mktemp -u)" 2>/dev/null \
+		| awk '$1 == "'"$conf"'" && /^[^ \t]/ { sub(/^[^ \t]+[ \t]+/, ""); print; exit }'
+	# match "datadir      /some/path with/spaces in/it here" but not "--xyz=abc\n     datadir (xyz)"
+}
+
+_start_server() {
+	local socket=$1; shift
+	"$@" --skip-networking --socket="${socket}" &
+	local pid="$!"
+
+	mysql=( mysql --protocol=socket -uroot -hlocalhost --socket="${socket}" )
+
+	for i in {30..0}; do
+		if echo 'SELECT 1' | "${mysql[@]}" &> /dev/null; then
+			break
+		fi
+		sleep 1
+	done
+	if [ "$i" = 0 ]; then
+		echo >&2 "$(date --rfc-3339=seconds) [ERROR] [Entrypoint]: Unable to start server."
+		exit 1
+	fi
+	return $pid
+}
+
+_stop_server() {
+	local server_pid="$1"
+	kill "$server_pid"
+	for i in $(seq 1 60); do
+		sleep 1
+		if ! $(pidof /usr/sbin/mysqld >/dev/null 2>&1); then
```

Comment:
Using `pidof` seems kind of hacky, especially if someone were to run this image with `--pid=host`; couldn't we instead use `--pid-file` (https://dev.mysql.com/doc/refman/8.0/en/mysql-server.html#option_mysql_server_pid-file) to have `mysqld` itself store this in a useful place for us to look up?

(Given that it has a reasonable default and doesn't appear to be disable-able, we might be able to get away with something like `_get_config 'pid-file'` to capture the appropriate value, but I'm not 100% sure where it'd be relative to if that is configured with a relative path?)

Reply:
Hm, yes. Normally just using the pid-file isn't enough (since it may have stale data, leading you to killing some other process instead).
And pid-file should never be a relative path :)

A couple of notes about starting/stopping the server, that we've implemented upstream and I might add to this PR:
1. 5.7+ Has the --daemonize option, which forks the process when ready, meaning no need to have a loop that checks if the server is up (https://github.com/mysql/mysql-docker/blob/mysql-server/5.7/docker-entrypoint.sh#L79)
2. When using a local socket, «mysqladmin shutdown» will block until the server has shut down (for 5.5+), so as long as the script has root access to the server, which it will during a fresh start (https://github.com/mysql/mysql-docker/blob/mysql-server/5.7/docker-entrypoint.sh#L160).

1. Is only a help for 5.7, but 2. should remove the need to pass the pid around